File: Javascript/merge.js
/**
 * @param {number[][]} intervals
 * @return {number[][]}
 */
var merge = function (intervals) {
    intervals.sort((a, b) => a[0] - b[0]);
    const result = [];
    let currentInt;
    for (let i = 0; i < intervals.length; i++) {
        if (i === 0) {
            currentInt = intervals[i];
        }
        else {
            const tmp = intervals[i];
            if (currentInt[0] <= tmp[0] && currentInt[1] >= tmp[0]) {
                if (currentInt[1] < tmp [1]) {
                    currentInt[1] = tmp[1];
                }
            }
            else {
                result.push(currentInt);
                currentInt = tmp;
            }
        }
    }
    result.push(currentInt);
    return result;
};

File: Python/merge.py
"""
leetcode 56. Merge Intervals
"""

class Solution:
    def merge(self, intervals: List[List[int]]) -> List[List[int]]:
        intervals.sort(key= lambda item: item[0])
        res = []
        for i in range(len(intervals)):
            if len(res) == 0:
                res.append(intervals[i].copy())
                continue
            begin = res[-1]
            end = intervals[i]
            if begin[0] == end[0] :
                if begin[1] < end[1]:
                    begin[1] = end[1]
            elif begin[1] < end[0]:
                res.append(end.copy())
            elif end[0] <= begin[1] <= end[1]:
                begin[1] = end[1]
        return res
